pypiserver/core.py

```diff
@@ -1,154 +1,68 @@
-    \.whl|\.dist-info)$""",
-    re.VERBOSE,
-)
-_pkgname_re = re.compile(r"-\d+[a-z_.!+]", re.I)
-_pkgname_parts_re = re.compile(
-    r"[\.\-](?=cp\d|py\d|macosx|linux|sunos|solaris|irix|aix|cygwin|win)", re.I
-)
-
-
-def _guess_pkgname_and_version_wheel(basename):
-    m = wheel_file_re.match(basename)
-    if not m:
-        return None, None
-    name = m.group("name")
-    ver = m.group("ver")
-    build = m.group("build")
-    if build:
-        return name, ver + "-" + build
-    else:
-        return name, ver
-
-
-def guess_pkgname_and_version(path):
-    path = os.path.basename(path)
-    if path.endswith(".asc"):
-        path = path.rstrip(".asc")
-    if path.endswith(".whl"):
-        return _guess_pkgname_and_version_wheel(path)
-    if not _archive_suffix_rx.search(path):
-        return
-    path = _archive_suffix_rx.sub("", path)
-    if "-" not in path:
-        pkgname, version = path, ""
-    elif path.count("-") == 1:
-        pkgname, version = path.split("-", 1)
-    elif "." not in path:
-        pkgname, version = path.rsplit("-", 1)
-    else:
-        pkgname = _pkgname_re.split(path)[0]
-        ver_spec = path[len(pkgname) + 1 :]
-        parts = _pkgname_parts_re.split(ver_spec)
-        version = parts[0]
-    return pkgname, version
-
-
-def normalize_pkgname(name):
-    """Perform PEP 503 normalization"""
-    return re.sub(r"[-_.]+", "-", name).lower()
-
-
-def normalize_pkgname_for_url(name):
-    """Perform PEP 503 normalization and ensure the value is safe for URLs."""
-    return quote(re.sub(r"[-_.]+", "-", name).lower())
-
-
-def is_allowed_path(path_part):
-    p = path_part.replace("\\", "/")
-    return not (p.startswith(".") or "/." in p)
+def get_bad_url_redirect_path(request, project):
+    """Get the path for a bad root url."""
+    uri = request.custom_fullpath
+    if uri.endswith("/"):
+        uri = uri[:-1]
+    uri = uri.rsplit("/", 1)[0]
+    project = quote(project)
+    uri += f"/simple/{project}/"
+    return uri
 
 
 class PkgFile:
-
     __slots__ = [
-        "fn",
-        "root",
-        "_fname_and_hash",
-        "relfn",
-        "relfn_unix",
-        "pkgname_norm",
-        "pkgname",
-        "version",
-        "parsed_version",
-        "replaces",
+        "pkgname",  # The projects/package name with possible capitalization
+        "version",  # The package version as a string
+        "fn",  # The full file path
+        "root",  # An optional root directory of the file
+        "relfn",  # The file path relative to the root
+        "replaces",  # The previous version of the package (used by manage.py)
+        "pkgname_norm",  # The PEP503 normalized project name
+        "digest",  # The file digest in the form of <algo>=<hash>
+        "relfn_unix",  # The relative file path in unix notation
+        "parsed_version",  # The package version as a tuple of parts
+        "digester",  # a function that calculates the digest for the package
```

Comment:
thanks for this documentation!

Reply:
Haha, I mostly put it there for my own sanity until I had the chance to refactor `PkgInfo`